src/modules/roc_core/target_bionic/roc_core/backtrace.cpp

```diff
@@ -6,15 +6,92 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
+#include <cxxabi.h>
+#include <iomanip>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <dlfcn.h>
+#include <unwind.h>
+
+
 #include "roc_core/backtrace.h"
 
+/* Implementation of backtrace in bionic/android is similar
+ * to the implementation of backtrace in glibc.
+ */
+struct BacktraceState
+{
+    void** current;
+    void** end;
+};
+
+static _Unwind_Reason_Code unwindCallback(struct _Unwind_Context* context, void* arg)
+{
+    BacktraceState* state = static_cast<BacktraceState*>(arg);
+    _Unwind_Ptr pc = _Unwind_GetIP(context);
+    if (pc) {
+	   if (state->current == state->end) {
+            return _URC_END_OF_STACK;
+       } else {
+            *state->current++ = reinterpret_cast<void*>(pc);
+       }
+    }
+    return _URC_NO_REASON;
+}
+
+size_t captureBacktrace(void** buffer, size_t max)
+{
+    BacktraceState state = {buffer, buffer + max};
+    _Unwind_Backtrace(unwindCallback, &state);
+
+    return state.current - buffer;
+}
+
+void dumpBacktrace(void** buffer, size_t count)
+{
+	if(count <= 0) {
+		fprintf(stderr, "No backtrace available\n");
+	} else {
+		fprintf(stderr, "Backtrace:\n");
+		for (size_t idx = 0; idx < count; ++idx) {
+			const void* addr = buffer[idx];
+			const char* symbol = "";
+			int status = -1;
+
+			Dl_info info;
+			if (dladdr(addr, &info) && info.dli_sname) {
+				symbol = info.dli_sname;
+			}
+			fprintf(stderr, "#%zd: 0x%x %s", idx, addr, symbol);
+			/* perform demangling
+			 */
+			symbol = abi::__cxa_demangle(symbol, 0, 0, &status);
```

Comment:
There is a leak here: __cxa_demangle() calls realloc() on its second argument (NULL in this case) and returns the result. So we should free() the demangled name.

https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a01696.html

Also, to reduce the number of allocations here, we can reuse the same buffer for demangling, and then free it at the end. You can find an example in the glibc version of backtrace.

Reply:
In glibc, the function name is copied into `mangled` array and demangled name is stored in `demangled`, where `mangled` is passed as a param to cxa_demangle(). So this part has to be implemented in bionic the same way?

Reply:
No. In the glibc version, the mangled name is copied to form a null-terminated string. Here you already have a null-terminated string (`symbol`) so there is no need to make a copy.

```
char* demangled_name = (char*)malloc(MaxLen * sizeof(char));
demangled_name = abi::__cxa_demangle(mangled_name, 0, 0, &status);
```

There is a leak here, again: you malloc() a buffer and then overwrite it with another buffer returned by __cxa_demangle(); so the original buffer will be never freed.

The correct and optimal (by number of allocations) code would be something like:

```
size_t demangled_size = MaxLen;
char* demangled_name = (char*)malloc(demangled_size);

for (...) {
   ...
    demangled_name =
        abi::__cxa_demangle(symbol, demangled_name, &demangled_size, &status);
    if (status == 0) {
        fprintf(stderr, " %s\n", demangled_name);
    } else {
        fprintf(stderr, " %s\n", symbol);
    }
}

free(demangled_name);
```

So you should pass demangled_name to __cxa_demangle() each time; and __cxa_demangle() will return either the same or reallocated buffer and will also update the demangled_size variable with the size of the buffer. When the loop finishes, you call free().